input/script_utils.go

```diff
@@ -1000,6 +1000,161 @@ func ReceiverHtlcSpendTimeout(signer Signer, signDesc *SignDescriptor,
+	cltvExpiry uint32) (txscript.TapLeaf, error) {
+
+	builder := txscript.NewScriptBuilder()
+
+	// The first part of the script will verify a signature from the
+	// receiver authorizing the spend.
+	builder.AddData(schnorr.SerializePubKey(receiverHtlcKey))
+	builder.AddOp(txscript.OP_CHECKSIG)
+	builder.AddOp(txscript.OP_CHECKSEQUENCEVERIFY)
+
+	// The second portion will ensure that the CLTV expiry on the spending
+	// transaction is correct.
+	builder.AddInt64(int64(cltvExpiry))
+	builder.AddOp(txscript.OP_CHECKLOCKTIMEVERIFY)
+	builder.AddOp(txscript.OP_DROP)
+
+	timeoutLeafScript, err := builder.Script()
+	if err != nil {
+		return txscript.TapLeaf{}, nil
+	}
+
+	return txscript.NewBaseTapLeaf(timeoutLeafScript), nil
+}
+
+// ReceiverHtlcTapLeafSuccess returns the full tapscript leaf for the success
+// path for an HTLC on the receiver's commitment transaction. This script
+// allows the receiver to redeem an HTLC with knowledge of the preimage:
+//
+// OP_SIZE 32 OP_EQUALVERIFY OP_HASH160
+// <RIPEMD160(payment_hash)> OP_EQUALVERIFY
+// <local_htlcpubkey> OP_CHECKSIGVERIFY
+// <remote_htlcpubkey> OP_CHECKSIG
+func ReceiverHtlcTapLeafSuccess(receiverHtlcKey *btcec.PublicKey,
+	senderHtlcKey *btcec.PublicKey,
+	paymentHash []byte) (txscript.TapLeaf, error) {
+
+	builder := txscript.NewScriptBuilder()
+
+	// Check that the pre-image is 32 bytes as required.
+	builder.AddOp(txscript.OP_SIZE)
+	builder.AddInt64(32)
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+
+	// Check that the specified pre-images matches what we hard code into
+	// the script.
+	builder.AddOp(txscript.OP_HASH160)
+	builder.AddData(Ripemd160H(paymentHash))
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+
+	// Verify the "2-of-2" multi-sig that requires both parties to sign
+	// off.
+	builder.AddData(schnorr.SerializePubKey(senderHtlcKey))
+	builder.AddOp(txscript.OP_CHECKSIGVERIFY)
+	builder.AddData(schnorr.SerializePubKey(receiverHtlcKey))
+	builder.AddOp(txscript.OP_CHECKSIG)
+
+	successLeafScript, err := builder.Script()
+	if err != nil {
+		return txscript.TapLeaf{}, err
+	}
+
+	return txscript.NewBaseTapLeaf(successLeafScript), nil
+}
+
+// receiverHtlcTapScriptTree builds the tapscript tree which is used to anchor
+// the HTLC key for HTLCs on the receiver's commitment.
+func receiverHtlcTapScriptTree(senderHtlcKey, receiverHtlcKey,
+	revokeKey *btcec.PublicKey, payHash []byte,
+	cltvExpiry uint32) (*HtlcScriptTree, error) {
+
+	// First, we'll obtain the tap leaves for both the success and timeout
+	// path.
+	successTapLeaf, err := ReceiverHtlcTapLeafSuccess(
+		receiverHtlcKey, senderHtlcKey, payHash,
+	)
+	if err != nil {
+		return nil, err
+	}
+	timeoutTapLeaf, err := ReceiverHtlcTapLeafTimeout(
+		receiverHtlcKey, cltvExpiry,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// With the two leaves obtained, we'll now make the tapscript tree,
+	// then obtain the root from that
+	tapscriptTree := txscript.AssembleTaprootScriptTree(
+		successTapLeaf, timeoutTapLeaf,
```

Comment:
So the ordering doesn't really matter here since the leaves are always sorted before hashing.

Reply:
oh i see. But I don't think the leaves are sorted inside `AssembleTaprootScriptTree` tho, made this test in `txscript/taproot_test.go`,
```go
func TestTapscriptLeafOrder(t *testing.T) {
	numLeafBytes := prand.Intn(1000)

	// Create leaf one.
	scriptBytes1 := make([]byte, numLeafBytes)
	_, err := prand.Read(scriptBytes1[:])
	require.NoError(t, err, "unable to read rand bytes")
	leaf1 := NewBaseTapLeaf(scriptBytes1)

	// Create leaf two.
	scriptBytes2 := make([]byte, numLeafBytes)
	_, err = prand.Read(scriptBytes2[:])
	require.NoError(t, err, "unable to read rand bytes")
	leaf2 := NewBaseTapLeaf(scriptBytes2)

	// Create tree with the order leaf1, leaf2.
	scriptTree12 := AssembleTaprootScriptTree(leaf1, leaf2)
	node1 := scriptTree12.RootNode

	// Create tree with the order leaf2, leaf1.
	scriptTree21 := AssembleTaprootScriptTree(leaf2, leaf1)
	node2 := scriptTree21.RootNode

	// Validate that the two trees are equal.
	require.Equal(t, node1.TapHash(), node2.TapHash())
	require.Equal(t, node1.Left().TapHash(), node2.Left().TapHash())
	require.Equal(t, node1.Right().TapHash(), node2.Right().TapHash())
}
```

Reply:
Correct the _leaves_ aren't sorted, but when you go to hash sorting happens. So this'll work as expected, since we only have two scripts. I modified the code to match up more directly though.